spec/thread_safety_spec.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+RSpec.describe "thread safety" do # rubocop:disable RSpec/DescribeClass
+  context "when two threads accessing unmemoized zero-args method" do
+    # Tip of the hat to @jeremyevans,
+    # see https://github.com/panorama-ed/memo_wise/pull/224
+    let(:thread_return_values) do
+      thread_return_values = [nil, nil]
+
+      1000.times do
+        instance = class_with_memo.new
+
+        threads = Array.new(2).map do
+          Thread.new { instance.return_first_thread_id }
+        end
+
+        thread_return_values = threads.map(&:value)
+
+        if thread_return_values.any?(&:nil?) ||
+           thread_return_values[0] != thread_return_values[1]
+          break
+        end
+      end
+
+      thread_return_values
+    end
+
+    let(:class_with_memo) do
+      Class.new do
+        prepend MemoWise
+
+        def return_first_thread_id
+          Thread.pass # trigger a race condition even on MRI
+          Thread.current.object_id
+        end
+        memo_wise :return_first_thread_id
+      end
+    end
+
+    # TODO: Pending proposed fix in #224, remove this comment and enable spec
+    pending "returns the same memoized non-nil thread id in each thread" do
+      expect(thread_return_values[0]).to eq(thread_return_values[1])
```

Comment:
I think this test will still fail even with @jeremyevans's fix. The rationale is that both threads will still hit this line:
```ruby
ret = @_memo_wise[#{index}] = #{original_memo_wised_name}
```
at roughly the same time, and still therefore get different values from their respective `return_first_thread_id` calls.

The case that Jeremy's PR has fixed is the nil case - neither will have a `nil` value. Do we also want to address this first case and ensure that both threads have the same memoized values? If so, might we need to use a Mutex or something like that (presumably at a performance cost in the non-memoized case) to lock the `@_memo_wise` cache?

Reply:
I think it's expected that any method that is cached should return the same value no matter which thread is calling it.  The test breaks that assumption.  Maybe the test should just check that the value returned is not nil, because in the failure case before my fix, `nil` should be returned if the race is lost.

Reply:
😝  I was hoping nobody would see that until we went to enable this test on Jeremy's PR

Reply:
(I was hoping to drive an explicit decision on this point within the context of his proposed fix)

What if I split the assertions into two, separate pending specs? And we can discuss whether Jeremy's PR should close one of them, and the other can be addressed in a separate PR, or converted to a documentary spec that shows that we expect possibly multiple values when the memoization is contended? (which to be clear is an acceptable design decision, that we simply need to document if so)

Reply:
@jeremyevans looks like we "typed past each other" here, as you can see in the timestamps of our comments above ;)

So to playback what we are all saying: this test has two distinct expectations.

Split-tracking:

- (a) Test of accidental `nil`
  - Proposal: Separate to its own test + remove `pending` as part of Jeremy's PR that fixes it?

- (b) Test of same return value when the memoization is contended
  - Option 1: Remove this test entirely
  - Option 2: Change test to be documentary - that this issue exists and should be expected
  - Option 3: Separate `pending` test - and plan to introduce a separate fix per Jeremy's description of the expectation above that we should not return different values when memoization is contended

Re-reading what everyone wrote above, what do we think about (a) separating, and (b-3) leaving a separate pending test in place to address the contended values issue in a different PR?

Reply:
I personally see no value in testing behavior for a non-idempotent method.  Nobody should be using caching for non-idempotent methods.  So I would only check that the race condition is fixed (the test for non-nil results on a method that returns non-nil).

Reply:
I am totally fine removing this case entirely for now. 

I suspect that, per #184, we must eventually document what happens in contended cases, for example:

- Queries of external state
  - Should reading data in from a web request, database query, or file be memoized?
  - If so, what is the actual vs expected behavior, when contended queries generate different query results?

Unless I hear back differently from @jemmaissroff or @JacobEvelyn, I will plan to remove this test case for now, and we can discuss under #184.

Reply:
Thanks for the thoughts, @ms-ati @jeremyevans @jemmaissroff ! I don't have strong feelings on what the behavior should be for nondeterministic methods, so I'm happy to defer to you three on either providing a stronger consistency guarantee or simply changing the test to confirm that the results are not `nil`. I agree with @ms-ati that we should document the behavior wherever we land.

Reply:
Per discussion above, I've removed the test of same return value under contended memoization, we can discuss this separately.